pyfr/backends/base/generator.py

```diff
@@ -132,69 +132,73 @@ def render(self):
 
     def _deref_arg_view(self, arg):
         ptns = [
-            '{0}_v[{0}_vix[X_IDX]]',
-            r'{0}_v[{0}_vix[X_IDX] + SOA_SZ*(\1)]',
-            r'{0}_v[{0}_vix[X_IDX] + {0}_vrstri[X_IDX]*(\1) + SOA_SZ*(\2)]'
+            '{0}_v[{0}_vix[{1}]]',
+            r'{0}_v[{0}_vix[{1}] + SOA_SZ*(\1)]',
+            r'{0}_v[{0}_vix[{1}] + {0}_vrstri[{1}]*(\1) + SOA_SZ*(\2)]'
         ]
 
-        return ptns[arg.ncdim].format(arg.name)
+        indx = 'ib*SZ + X_IDX' if True else 'X_IDX'
```

Comment:
I just left this because I couldn't find a way to check the backend.cacheblks from this class. Any suggestions?

Reply:
You can start passing it to the constructor.

Reply:
Could you elaborate on this?

Reply:
So you can pass the cachblks parameter to the constructor of the generator.  However, the usual solution here is to keep the argument as X_IDX and then #define it appropriately in the sub class.  This is how we currently deal with the fact that it has a different definition for OpenMP than CUDA.

Reply:
But cacheblks parameter is defined in OpenMPBackend class. generator stuff goes through a different path. I don't understand how this can be done.

For deref_arg_view, index doesn't change based on the arg attribute so defining later would work. But for 1d and 2d deref funcs when we are dealing with matrices, the amount of offset per block is a function of matrix blocksz and this requires knowing ndim and nvar. And obviously it is not a common value for all the matrices in a kernel. Just like the current dereferencing procedure, it is sensible to come back to this function to determine these attributes based on the particular matrix. That's why I implemented in this way. 

An alternative would be passing blocksz of each array to the c function just like we pass ldim for each. But then, it is a variable and just like ldim we have to include it somehow in the deref function's 'ix' because it is a different value for each matrix. A common #define for all matrices wouldn't work here. It kind of comes back to what I did, and mine is simpler because we don't need to pass anything extra to kernels.

Reply:
So it can be passed into the class in much the same way we pass fpdtype at the moment; see https://github.com/PyFR/PyFR/blob/develop/pyfr/backends/base/makoutil.py#L143

However, I am still not convinced it is needed.  Rather it would be better to always generate blocking indices and then let them be zeroed later on by a #define.  For example we can use BLK_IDX in place of ib and then backends which do not use blocking can `#define BLK_IDX 0` and its contribution gets zeroed.  OpenMP can then `#define BLK_IDX ib` and have everything work as-is.